merge main into release/7.0-preview5 branch

smasher164 opened the pull request:
The snap for preview 5 is happening at 2PM today, but this branch had been created in the morning.
This PR incorporates the commits since then into the release branch.

radekdoulik committed:
[wasm] retrieve results from browser bench sample (#69673)

Add simple http server to serve files for the browser-bench sample and let it collect the results at the end.

The results are sent as POST requests and the simple server saves them to the host storage.

* Save the results with POST requests

* Add simple http server

It also saves POST requests and is used to collect results
of browser-bench sample

* Let RunSample target use the simple server

MihaZupan committed:
Apply MaxResponseHeadersLength limit to trailing headers (#69250)

* Apply MaxResponseHeadersLength limit to trailing headers

* PR feedback

* Drop 'envelope' from the error message

LakshanF committed:
Fix Repro to work with recent changes (#69769)

* Fix Repro to work with recent changes

* FB

shushanhf committed:
[LoongArch64] Fix some errors for LA-ABI. (#69771) (#69771)

tmds committed:
OpenSslX509ChainProcessor: ignore NotSignatureValid if it occurs on a trusted root

The new RHEL/CentOS/Fedora RSA+SHA-1 policy causes us to report all chains ending in an RSA+SHA-1 root as invalid.

Since OpenSSL doesn't normally check signatures on self-issued certs, but we ask it to (for Windows/NetFX compat), we fail when tools like `curl` succeed.

So now we'll still ask OpenSSL to do the work, but if the NotSignatureValid is only on the last element, and we didn't get UntrustedRoot or PartialChain, then we ignore it and say that everything's copacetic.

maraf committed:
[wasm] Consolidate interop tables generators (#69620)

Combine PInvoke, icall and ManagedToNative generators to a single MSBuild task.

steveisok committed:
[iOS] Enable LLVM on device runs (#69504)

This change makes sure iOS & tvOS device runs enable LLVM mode

monojenkins committed:
switch domain and thread TLS clearing in mono_thread_detach_internal (#69796)

Backported from https://github.com/mono/mono/pull/21482

* Set app context to null on thread detach

When a domain is unloaded, it is possible that an InternalThread
has a reference to an app context from the unloaded domain.

* There is no SET_APPCONTEXT in netcore

So the substantive portion of the mono/mono PR is gone.  This just changes around the order in which we clear the domain and thread object TLS references.

Co-authored-by: bholmes <bholmes@users.noreply.github.com>
Co-authored-by: Aleksey Kliger (λgeek) <akliger@gmail.com>

veanes committed:
Improve regex derivative construction (#69530)

* testing some ideas

* derivative computation updates

* Capture support with new derivative and fixes

Introduce new Effect node type that derivatives can include.
Logic around derivative functions to translate Effect nodes to
DerivativeEffect or just strip them away.
Remove old AddTransitions derivative logic.
Misc fixes to NFA support that was still assuming old unordered Or.

* Fix StripEffects and improve OrderedOr constructor

StripEffects now avoids repeated deduplication.
The OrderedOr in the builder now does deduplication efficiently.

* Add nothing rule for Effect nodes

* Generic subsumption check and rule in OrderedOr

* Got things to a state where all stress tests pass

* Strip out currently unused stuff and refactor

* Revert "Strip out currently unused stuff and refactor"

This reverts commit 11a2cfdad26e32a02ad1558282f64ed4a247e02d.

* Subsume and fold rule

* Make the subsumption fast by reordering rules

* added caching of pruning

* Small fix

* added comments

* Cleanup

* Fix bug introduced with cleanup in fold

Also remove concat loop rules

* More cleanup

* More cleanup

* Cleanup and revert some upcoming changes

* renaming ToString to ToStringHelper to avoid build break

* avoid overloading ToString

* forgotten if DEBUG inserted back

* Add lazy concat stress test and optimizations

* Comments

* Comments and a fix

* More comments

* Fix to ToString to match preferred Effect notation

* Add longer comments about Effects

* More comments and a code simplification

* Apply suggestions from code review

Co-authored-by: Stephen Toub <stoub@microsoft.com>

* Syntax fix

* fixed remaining comments

Co-authored-by: Olli Saarikivi <olsaarik@microsoft.com>
Co-authored-by: Stephen Toub <stoub@microsoft.com>

jkotas committed:
Fold System.Private.Interop into CoreLib (#69508)

kg committed:
Reimplement I52/U52 in C, clean up and optimize range checks (#69624)

This PR reimplements the I52/U52 memory accessors in C since the JS got really complicated and we kept finding bugs. It also simplifies some of the range checks in the other accessors so they will have less overhead, and uses some unchecked accessors in a few hot paths.
- Move _zero_region to memory and export it. Zero the stack-allocated buffer used by converters

radical committed:
[wasm] First attempt at a wasm app host (#68696)

First attempt at implementing a wasm app host

Regular dotnet apps can be run with `dotnet run`. But this is not the case for
wasm apps. We use `xharness` to run the test apps that we generate for library
tests.

Templates for browser, and console wasm projects were added recently, but they
have to be run manually by either invoking a v8 script, or running a
`http-server` in the bundle directory. Though to use a debugger with this
requires starting the debug proxy separately, and connecting few bits.

# WasmAppHost

This PR adds a new `WasmAppHost`, which can be used through `dotnet run`:

- it can run "wasm app" from it's `AppBundle` directory, using various hosts
  like `v8`, `node`, or a browser. An example use:

    ```
    $ dotnet new wasmconsole
    $ dotnet run
    WasmAppHost --runtime-config /tmp/c0/bin/Debug/net7.0/browser-wasm/AppBundle//c0.runtimeconfig.json
    [02:26:40] info: host[0] Running: node main.cjs
    [02:26:40] dbug: host[0] Using working directory: /tmp/c0/bin/Debug/net7.0/browser-wasm/AppBundle
    Incoming arguments:
    Application arguments:
    Debugger.Debug: DEBUGGING ENABLED

    mono_wasm_runtime_ready fe00e07a-5519-4dfe-b35a-f867dbaf2e28
    Could not find symbols file dotnet.js.symbols. Ignoring.
    Initializing.....
    Hello, Console!
    ```

    The console template project is run with `node` by default.

- it supports a `--debug` parameter, which (for the browser case) would startup
  a webserver, and the debug proxy for Chrome, and firefox.

```
$ dotnet new wasmbrowser

$ dotnet run --debug
WasmAppHost --runtime-config /workspaces/test/br0/bin/Debug/net7.0/browser-wasm/AppBundle//br0.runtimeconfig.json --debug --forward-console
Debug proxy for chrome now listening on http://127.0.0.1:9300/. And expecting chrome at http://localhost:9222/
Hosting environment: Production
Content root path: /workspaces/test/br0/bin/Debug/net7.0/browser-wasm/AppBundle
Now listening on: http://127.0.0.1:9300
Debug proxy for firefox now listening on tcp://127.0.0.1:6300. And expecting firefox at port 6000 .

App url: http://127.0.0.1:9000/index.html
App url: https://127.0.0.1:38331/index.html
```

- This enables using `dotnet run` with wasm samples, and library tests too.

## How do I try this out?

You'll need a dotnet with the updated packs. For now, you can use:

```
$ ./dotnet.sh build -p:TargetOS=Browser -p:TargetArchitecture=wasm -p:Configuration=Release src/tests/BuildWasmApps/Wasm.Build.Tests/Wasm.Build.Tests.csproj /t:InstallWorkloadUsingArtifacts
```

.. which would install a sdk in `artifacts/bin/dotnet-workload` with all the packs installed.

- Use that dotnet to create console, or browser wasm apps from templates, for example `dotnet new wasmconsole`, or `dotnet new wasmbrowser`.

### Debugging library tests

```
runtime$ ./artifacts/bin/dotnet-workload/dotnet run -f net7.0 -r browser-wasm -c $YourRuntimeConfig --project src/libraries/System.Buffers/tests/System.Buffers.Tests.csproj -p:DebuggerSupport=true --debug --host browser
```

- If the runtime was built in `Release` config, then you need to add `-p:DebuggerSupport=true`, so the pdbs would get deployed
- And use `--debug` to start the debugger

- Opening the url will show a `Run tests` button, which you can click after connecting with the IDE, to start running the tests

## Notes, and TODO

- `WasmAppHost`, along with the debug proxy are bundled in the `WebAssembly.Sdk` pack right now.
  - This might be changed in the future to use the dotnet app host pack pattern, and with
    additional support adding in the Sdk.
  - This would also allow using `launchSettings.json`, similar to blazor

- `WasmAppHost` depends on host config in `runtimeconfig.json`, which can be
  controlled by a `runtimeconfig.template.json` in the project directory.

- Needs more tests, cleanup, `--help` for command line arguments
- `app-support.*js` in templates, and `test-main.js` need more work to share the code

Fixes https://github.com/dotnet/runtime/issues/67236

AaronRobinsonMSFT committed:
Revert MSVC changes for libunwind (#69750)

* Revert MSVC changes to libunwind

AaronRobinsonMSFT committed:
Fix memory issues detected with GCC 12.1 (#69778)

* Fix memory issues detected with GCC 12

elinor-fung committed:
[LibraryImportGenerator] Allow span copy for char arrays instead of manual copy loop (#69764)

carlossanlop reviewed (approved):
LGTM, the commits match those in main.

Will merge this with a "merge commit".

BrzVlad committed:
[mono][interp] Add tiering within interpreter (#68823)

* [mono][interp] Remove error argument from mono_interp_get_imethod

It is unused

* [mono][interp] Implement tiering at method execution start

For a single MonoMethod*, we can have two InterpMethod* instances, one with optimized flag false and the other with true. When tiering is enabled, when first getting an InterpMethod* for a MonoMethod* we set the optimized flag to false. When generatig code for this method, if optimized is false we must emit a special MINT_TIER_ENTER_METHOD at the start and later in the codegen process we skip applying optimizations to method code.

MINT_TIER_ENTER_METHOD opcode is invoked with every method start and it will bump a counter. Once we hit the limit, the method will be tiered up. This process consists of creating a new InterpMethod* instance which have optimized set and storing it in the interp_code_hash, changing the mapping from the old MonoMethod. The optimized and unoptimized method use the same argument space, so tiering the method up requires just to set the ip to the start of the tiered up method code.

An additional problem that happens with tiering is that we have to replace all instances of the untiered method from generated code. InterpMethod* instances are stored stored inside data_items of other methods and also inside vtables. When generating code for any method, we have to store in a hash table mappings from untiered InterpMethod* instance to addresses where this instance was stored. When we tier up the unoptimized method, we will traverse the list of addresses where this references is stored and update it to the optimized version.

* [mono][interp] Add option to enable/disable optimizations

Some optimizations might not be enabled by default, so add option to enable them.

* [mono][interp] Implement on stack replacement tiering up

In unoptimized code, we add a patchpoint instruction when entering basic blocks that are targets of backward branches, if the stack state is empty. This means that when tiering up a frame we just need to jump to the equivalent basic block in the tiered up method and execution can continue. Since the arguments and IL locals reside in the same space in both versions of the method (their offsets are computed in interp_method_compute_offsets)

* [mono][interp] Remove calc_section mutex

We always take jit_mm lock when finishing compilation of method, use it also for publishing InterpMethod* fields. This also prevents weird races where the method can be tiered up before the we take the jit_mm lock, resulting in publishing the seq_points for the untiered method

* [mono][interp] Fix incorrect linking of bblocks

Once we emit a tailcall, execution in the current bblock is finished.

* [mono][interp] Fix implicit conversion from i4 to native int

We were doing unsigned conversion before

* [mono][interp] Fix execution of clauses from il state when tiering is enabled

When invoking these clauses we obtained the InterpMethod from the MonoMethod* and make use of the jit info stored during frame unwinding. However, the method might have been tiered up since storing the jit info, so the native offsets stored there will no longer be relative to the optimized imethod. Fetch again the MonoJitInfo* from the imethod that we will be executing.

* [mono][interp] Enable tiering by default

vcsjones committed:
Use stack buffers and return data to pool in EC point encoding.
